File: leetcode/cpp/PassedProblems/205-isomorphic-strings.cpp
/*
 * 205-isomorphic-strings.cpp
 *
 *  Created on: Sep 14, 2015
 *      Author: hanzhe
 */
#include <iostream>
#include <algorithm>
using namespace std;
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        int * sa = new int[255], * sb = new int[255];
        for(int i = 0 ; i < 255 ; i ++)
        	sa[i] = sb[i] = 0;
        for(int i = 0 ; i < s.length(); i ++)
        {
        	if(sa[s[i] - 0] == 0  && sb[t[i]] == 0){
        		sa[s[i] - 0] = t[i] - 0;
        		sb[t[i]] = s[i];
        	}
        	else if(sa[s[i]] != t[i] || sb[t[i]] != s[i])
        		return false;
        }
        return true;
    }

    void test(){
    	cout << isIsomorphic("ab", "aa") << endl;
    	cout << isIsomorphic("egg", "add") << endl;
    	cout << isIsomorphic("foo", "bar") << endl;
    	cout << isIsomorphic("paper", "title") << endl;
    }
};

File: leetcode/python/205-isomorphic-strings.py
#!/usr/bin/env python
# encoding: utf-8

"""
Given two strings s and t, determine if they are isomorphic.

Two strings are isomorphic if the characters in s can be replaced to get t.

All occurrences of a character must be replaced with another character while preserving the order of characters. No two characters may map to the same character but a character may map to itself.

For example,
Given "egg", "add", return true.

Given "foo", "bar", return false.

Given "paper", "title", return true.

Note:
You may assume both s and t have the same length.
"""


# 2次过，return是二维数组，不是一维数组
# python可以判断二维int数组只是否相等
class Solution(object):
    def isIsomorphic(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        def lcmp(l1, l2):
            return 1 if l1[0] > l2[0] else -1
        def s2l(str):
            dic = {}
            for i in range(len(str)):
                c = str[i]
                if c not in dic:
                    dic[c] = []
                dic[c].append(i)
            ls = []
            for x in dic:
                ls.append(dic[x])
            ls.sort(lcmp)
            return ls
        return s2l(s) == s2l(t)

